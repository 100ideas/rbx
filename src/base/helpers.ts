import classNames from "classnames";
import * as PropTypes from "prop-types";

import { Omit, Prefer } from "../types";
import { tuple } from "../utils";

/**
 * Float
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export const FLOAT_PULLED_ALIGNMENTS = tuple("left", "right");
export type FloatPulledAlignments = (typeof FLOAT_PULLED_ALIGNMENTS)[number];

export type FloatHelpersProps = Partial<{
  /** Fixes an element's floating children */
  clearfix: boolean;
  /** Moves an element to the left or right */
  pull: FloatPulledAlignments;
}>;

export const floatHelpersPropTypes = {
  clearfix: PropTypes.bool,
  pull: PropTypes.oneOf(FLOAT_PULLED_ALIGNMENTS),
};

export const transformFloatHelpers: TransformFunc<FloatHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    floatHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, clearfix, pull, ...rest } = props;

  return {
    className: classNames(
      {
        "is-clearfix": clearfix,
        [`is-pulled-${pull}`]: pull,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Overflow
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export type OverflowHelpersProps = Partial<{
  /** Adds overflow hidden */
  clipped: boolean;
}>;

export const overflowHelpersPropTypes = {
  clipped: PropTypes.bool,
};

export const transformOverflowHelpers: TransformFunc<OverflowHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    overflowHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, clipped, ...rest } = props;

  return {
    className: classNames({ "is-clipped": clipped }, className),
    ...rest,
  };
};

/**
 * Overlay
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export type OverlayHelpersProps = Partial<{
  /** Completely covers the first positioned parent */
  overlay: boolean;
}>;

export const overlayHelpersPropTypes = {
  overlay: PropTypes.bool,
};

export const transformOverlayHelpers: TransformFunc<OverlayHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    overlayHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, overlay, ...rest } = props;

  return {
    className: classNames({ "is-overlay": overlay }, className),
    ...rest,
  };
};

/**
 * Typography
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export const COLORS = tuple(
  "primary",
  "success",
  "info",
  "warning",
  "danger",
  "light",
  "dark",
  "white",
  "black",
  "link",
);
export type Colors = (typeof COLORS)[number];

export const GREY_COLORS = tuple(
  "black-bis",
  "black-ter",
  "grey-darker",
  "grey-dark",
  "grey",
  "grey-light",
  "grey-lighter",
  "white-ter",
  "white-bis",
);
export type GreyColors = (typeof GREY_COLORS)[number];

export const TEXT_ALIGNMENTS = tuple("centered", "justified", "left", "right");
export type TextAlignments = (typeof TEXT_ALIGNMENTS)[number];

export const TEXT_SIZES = tuple(1, 2, 3, 4, 5, 6);
export type TextSizes = (typeof TEXT_SIZES)[number];

export const TEXT_TRANSFORMS = tuple("capitalized", "lowercase", "uppercase");
export type TextTransforms = (typeof TEXT_TRANSFORMS)[number];

export const TEXT_WEIGHTS = tuple("light", "normal", "semibold", "bold");
export type TextWeights = (typeof TEXT_WEIGHTS)[number];

export type TypographyHelpersProps = Partial<{
  backgroundColor: Colors | GreyColors;
  italic: boolean;
  textAlignment: TextAlignments;
  textColor: Colors | GreyColors;
  textSize: TextSizes;
  textTransform: TextTransforms;
  textWeight: TextWeights;
}>;

export const typographyHelpersPropTypes = {
  backgroundColor: PropTypes.oneOf([...COLORS, ...GREY_COLORS]),
  italic: PropTypes.bool,
  textAlignment: PropTypes.oneOf(TEXT_ALIGNMENTS),
  textColor: PropTypes.oneOf([...COLORS, ...GREY_COLORS]),
  textSize: PropTypes.oneOf(TEXT_SIZES),
  textTransform: PropTypes.oneOf(TEXT_TRANSFORMS),
  textWeight: PropTypes.oneOf(TEXT_WEIGHTS),
};

export const transformTypographyHelpers: TransformFunc<
  TypographyHelpersProps
> = (props, componentName, location = "prop") => {
  PropTypes.checkPropTypes(
    typographyHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const {
    backgroundColor,
    className,
    italic,
    textAlignment,
    textColor,
    textSize,
    textTransform,
    textWeight,
    ...rest
  } = props;

  return {
    className: classNames(
      {
        [`has-background-${backgroundColor}`]: backgroundColor,
        [`has-text-${textColor}`]: textColor,
        "is-italic": italic,
        [`is-${textTransform}`]: textTransform,
        [`has-text-${textAlignment}`]: textAlignment,
        [`has-text-weight-${textWeight}`]: textWeight,
        [`is-size-${textSize}`]: textSize,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Visibility
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export const DISPLAYS = tuple(
  "block",
  "flex",
  "inline",
  "inline-block",
  "inline-flex",
);
export type Displays = (typeof DISPLAYS)[number];

export type VisibilityHelpersProps = Partial<{
  /** Hides an element (unclear on where this is documented in Bulma) */
  hidden: boolean;
  /** Adds visibility hidden */
  invisible: boolean;
  /**
   * Hide elements visually but keep the element available to be announced by a
   * screen reader
   */
  srOnly: boolean;
}>;

export const visibilityHelpersPropTypes = {
  hidden: PropTypes.bool,
  invisible: PropTypes.bool,
  srOnly: PropTypes.bool,
};

export const transformVisibilityHelpers: TransformFunc<
  VisibilityHelpersProps
> = (props, componentName, location = "prop") => {
  PropTypes.checkPropTypes(
    visibilityHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, hidden, invisible, srOnly, ...rest } = props;

  return {
    className: classNames(
      {
        "is-hidden": hidden,
        "is-invisible": invisible,
        "is-sr-only": srOnly,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Other
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export type OtherHelpersProps = Partial<{
  /** Removes any margin */
  marginless: boolean;
  /** Removes any padding */
  paddingless: boolean;
  /** Removes any radius */
  radiusless: boolean;
  /** Removes any shadow */
  shadowless: boolean;
  /** Prevents the text from being selectable */
  unselectable: boolean;
}>;

export const otherHelpersPropTypes = {
  marginless: PropTypes.bool,
  paddingless: PropTypes.bool,
  radiusless: PropTypes.bool,
  shadowless: PropTypes.bool,
  unselectable: PropTypes.bool,
};

export const transformOtherHelpers: TransformFunc<OtherHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    otherHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const {
    className,
    marginless,
    paddingless,
    radiusless,
    shadowless,
    unselectable,
    ...rest
  } = props;

  return {
    className: classNames(
      {
        "is-marginless": marginless,
        "is-paddingless": paddingless,
        "is-radiusless": radiusless,
        "is-shadowless": shadowless,
        "is-unselectable": unselectable,
      },
      className,
    ),
    ...rest,
  };
};

/**
 * Responsive
 */
export const BREAKPOINTS = tuple(
  "mobile",
  "tablet",
  "desktop",
  "widescreen",
  "fullhd",
  "touch",
);
export type Breakpoints = (typeof BREAKPOINTS)[number];

export type LimitiedResponsiveBreakpointProps = Partial<{
  display: {
    value: Displays;
  };
  hide: {
    value: boolean;
  };
  textAlignment: {
    value: TextAlignments;
  };
  textSize: {
    value: TextSizes;
  };
}>;

export type ResponsiveBreakpointProps = Partial<{
  display: {
    only?: boolean;
    value: Displays;
  };
  hide: {
    only?: boolean;
    value: boolean;
  };
  textAlignment: {
    only?: boolean;
    value: TextAlignments;
  };
  textSize: {
    only?: boolean;
    value: TextSizes;
  };
}>;

export type ResponsiveHelpersProps = Partial<{
  responsive: Partial<{
    mobile: LimitiedResponsiveBreakpointProps;
    tablet: ResponsiveBreakpointProps;
    desktop: ResponsiveBreakpointProps;
    widescreen: ResponsiveBreakpointProps;
    fullhd: LimitiedResponsiveBreakpointProps;
    touch: LimitiedResponsiveBreakpointProps;
  }>;
}>;

export const responsiveBreakpointPropTypes = {
  display: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.oneOf(DISPLAYS).isRequired,
  }),
  hide: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.bool.isRequired,
  }),
  textAlignment: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.oneOf(TEXT_ALIGNMENTS).isRequired,
  }),
  textSize: PropTypes.shape({
    only: PropTypes.bool,
    value: PropTypes.oneOf(TEXT_SIZES).isRequired,
  }),
};

export const limitedResponsiveBreakpointPropTypes = {
  display: PropTypes.shape({
    value: PropTypes.oneOf(DISPLAYS).isRequired,
  }),
  hide: PropTypes.shape({
    value: PropTypes.bool.isRequired,
  }),
  textAlignment: PropTypes.shape({
    value: PropTypes.oneOf(TEXT_ALIGNMENTS).isRequired,
  }),
  textSize: PropTypes.shape({
    value: PropTypes.oneOf(TEXT_SIZES).isRequired,
  }),
};

export const responsiveHelpersPropTypes = {
  responsive: PropTypes.shape({
    // tslint:disable:object-literal-sort-keys
    mobile: PropTypes.shape(limitedResponsiveBreakpointPropTypes),
    tablet: PropTypes.shape(responsiveBreakpointPropTypes),
    desktop: PropTypes.shape(responsiveBreakpointPropTypes),
    widescreen: PropTypes.shape(responsiveBreakpointPropTypes),
    fullhd: PropTypes.shape(limitedResponsiveBreakpointPropTypes),
    touch: PropTypes.shape(limitedResponsiveBreakpointPropTypes),
    // tslint:enable:object-literal-sort-keys
  }),
};

export const transformResponsiveHelpers: TransformFunc<
  ResponsiveHelpersProps
> = (props, componentName, location = "prop") => {
  PropTypes.checkPropTypes(
    responsiveHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, responsive, ...rest } = props;

  return {
    className: classNames(
      responsive !== undefined
        ? Object.keys(responsive)
            .filter(breakpoint => responsive[breakpoint] !== undefined)
            .map(breakpoint => {
              const names = {};
              const { display, hide, textAlignment, textSize } = responsive[
                breakpoint
              ] as
                | ResponsiveBreakpointProps
                | LimitiedResponsiveBreakpointProps;
              if (display !== undefined) {
                const value = display.value;
                const only = "only" in display ? display.only === true : false;
                names[
                  `is-${value}-${breakpoint}${only ? "-only" : ""}`
                ] = value;
              }
              if (hide !== undefined) {
                const value = hide.value;
                const only = "only" in hide ? hide.only === true : false;
                names[`is-hidden-${breakpoint}${only ? "-only" : ""}`] = value;
              }
              if (textAlignment !== undefined) {
                const value = textAlignment.value;
                const only =
                  "only" in textAlignment ? textAlignment.only === true : false;
                names[
                  `has-text-${value}-${breakpoint}${only ? "-only" : ""}`
                ] = value;
              }
              if (textSize !== undefined) {
                const value = textSize.value;
                const only =
                  "only" in textSize ? textSize.only === true : false;
                names[
                  `is-size-${value}-${breakpoint}${only ? "-only" : ""}`
                ] = value;
              }

              return names;
            })
            .reduce((acc, cv) => ({ ...acc, ...cv }), {})
        : undefined,
      className,
    ),
    ...rest,
  };
};

/**
 * Union of helpers
 */
// tslint:disable-next-line: no-empty-interface
export interface HelpersPropsOverrides {}

export type HelpersProps = Prefer<
  HelpersPropsOverrides,
  FloatHelpersProps &
    OverflowHelpersProps &
    OverlayHelpersProps &
    TypographyHelpersProps &
    VisibilityHelpersProps &
    OtherHelpersProps &
    ResponsiveHelpersProps & { className?: string }
>;

export const helpersPropTypes = {
  ...floatHelpersPropTypes,
  ...overflowHelpersPropTypes,
  ...overlayHelpersPropTypes,
  ...typographyHelpersPropTypes,
  ...visibilityHelpersPropTypes,
  ...otherHelpersPropTypes,
  ...responsiveHelpersPropTypes,
  className: PropTypes.string,
};

export const combineTransformFunctions = <TTransformProps>(
  ...funcs: TransformFunc<any>[] // tslint:disable-line:no-any
): TransformFunc<TTransformProps> => (
  props,
  componentName,
  location = "prop",
) =>
  // tslint:disable-next-line:no-any
  funcs.reduce((acc, func) => func(acc, componentName, location) as any, props);

export const transformHelpers = combineTransformFunctions<HelpersProps>(
  transformFloatHelpers,
  transformOverflowHelpers,
  transformOverlayHelpers,
  transformTypographyHelpers,
  transformVisibilityHelpers,
  transformOtherHelpers,
  transformResponsiveHelpers,
);

export type TransformFunc<TTransformProps extends {}> = <
  TProps extends TTransformProps & { className?: string }
>(
  props: TProps,
  componentName: string,
  location?: string,
) => Omit<TProps, keyof TTransformProps | "className"> & { className?: string };
