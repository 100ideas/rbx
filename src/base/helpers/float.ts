import classNames from "classnames";
import * as PropTypes from "prop-types";

import { tuple } from "../../utils";
import { TransformFunc } from "./types";

/**
 * Float
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export const FLOAT_PULLED_ALIGNMENTS = tuple("left", "right");
export type FloatPulledAlignments = (typeof FLOAT_PULLED_ALIGNMENTS)[number];

export type FloatHelpersProps = Partial<{
  /** Fixes an element's floating children */
  clearfix: boolean;
  /** Moves an element to the left or right */
  pull: FloatPulledAlignments;
}>;

export const makeFloatHelpersPropTypesDefaults = {
  floatPulledAssignments: FLOAT_PULLED_ALIGNMENTS,
};

export const makeFloatHelpersPropTypes = (
  options: {
    floatPulledAssignments?: string[];
  } = makeFloatHelpersPropTypesDefaults,
) => {
  const values = { ...options, ...makeFloatHelpersPropTypesDefaults };

  return {
    clearfix: PropTypes.bool,
    pull: PropTypes.oneOf(values.floatPulledAssignments),
  };
};

export const floatHelpersPropTypes = makeFloatHelpersPropTypes();

export const transformFloatHelpers: TransformFunc<FloatHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    floatHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const { className, clearfix, pull, ...rest } = props;

  return {
    className: classNames(
      {
        "is-clearfix": clearfix,
        [`is-pulled-${pull}`]: pull,
      },
      className,
    ),
    ...rest,
  };
};
