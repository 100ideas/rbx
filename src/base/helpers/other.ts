import classNames from "classnames";
import * as PropTypes from "prop-types";

import { TransformFunc } from "./types";

/**
 * Other
 * https://github.com/jgthms/bulma/blob/master/sass/base/helpers.sass
 */
export type OtherHelpersProps = Partial<{
  /** Removes any margin */
  marginless: boolean;
  /** Removes any padding */
  paddingless: boolean;
  /** Removes any radius */
  radiusless: boolean;
  /** Removes any shadow */
  shadowless: boolean;
  /** Prevents the text from being selectable */
  unselectable: boolean;
}>;

export const otherHelpersPropTypes = {
  marginless: PropTypes.bool,
  paddingless: PropTypes.bool,
  radiusless: PropTypes.bool,
  shadowless: PropTypes.bool,
  unselectable: PropTypes.bool,
};

export const transformOtherHelpers: TransformFunc<OtherHelpersProps> = (
  props,
  componentName,
  location = "prop",
) => {
  PropTypes.checkPropTypes(
    otherHelpersPropTypes,
    props,
    location,
    componentName,
  );
  const {
    className,
    marginless,
    paddingless,
    radiusless,
    shadowless,
    unselectable,
    ...rest
  } = props;

  return {
    className: classNames(
      {
        "is-marginless": marginless,
        "is-paddingless": paddingless,
        "is-radiusless": radiusless,
        "is-shadowless": shadowless,
        "is-unselectable": unselectable,
      },
      className,
    ),
    ...rest,
  };
};
