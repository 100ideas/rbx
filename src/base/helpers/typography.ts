import classNames from "classnames";
import PropTypes from "prop-types";

import {
  makePropTypesFactory,
  makeValidatingTransformFactory,
  TransformFunction,
} from "./factory";
import { Variables } from "./variables";

export type TypographyHelpersProps = Partial<{
  backgroundColor: Variables["colors"] | Variables["shades"];
  italic: boolean;
  textAlignment: Variables["textAlignments"];
  textColor: Variables["colors"] | Variables["shades"];
  textSize: Variables["textSizes"];
  textTransform: Variables["textTransforms"];
  textWeight: Variables["textWeights"];
}>;

// Factories
export const makePropTypes = makePropTypesFactory(vars => ({
  backgroundColor: PropTypes.oneOf([...vars.colors, ...vars.shades]),
  italic: PropTypes.bool,
  textAlignment: PropTypes.oneOf(vars.textAlignments),
  textColor: PropTypes.oneOf([...vars.colors, ...vars.shades]),
  textSize: PropTypes.oneOf(vars.textSizes),
  textTransform: PropTypes.oneOf(vars.textTransforms),
  textWeight: PropTypes.oneOf(vars.textWeights),
}));

export const transform: TransformFunction<TypographyHelpersProps> = props => {
  const {
    backgroundColor,
    italic,
    textAlignment,
    textColor,
    textSize,
    textTransform,
    textWeight,
    ...rest
  } = props;

  // Can remove "no-any" and "no-unsafe-any" with TypeScript 3.3
  // https://github.com/Microsoft/TypeScript/pull/29121
  // tslint:disable:no-any
  // tslint:disable:no-unsafe-any
  (rest as any).className = classNames(
    {
      [`has-background-${backgroundColor}`]: backgroundColor,
      [`has-text-${textColor}`]: textColor,
      "is-italic": italic,
      [`is-${textTransform}`]: textTransform,
      [`has-text-${textAlignment}`]: textAlignment,
      [`has-text-weight-${textWeight}`]: textWeight,
      [`is-size-${textSize}`]: textSize,
    },
    (rest as any).className,
  );

  return rest;
};

export const makeValidatingTransform = makeValidatingTransformFactory(
  makePropTypes,
  transform,
);
