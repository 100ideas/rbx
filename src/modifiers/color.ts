import { tuple } from "@/utils";
import { cx } from "emotion";

import { makeTransform } from "./utils";

export const COLORS = tuple(
  "primary",
  "success",
  "info",
  "warning",
  "danger",
  "light",
  "dark",
  "white",
  "black",
  "link",
);
export type Colors = (typeof COLORS)[number];

export const GREY_COLORS = tuple(
  "black-bis",
  "black-ter",
  "grey-darker",
  "grey-dark",
  "grey",
  "grey-light",
  "grey-lighter",
  "white-ter",
  "white-bis",
);
export type GreyColors = (typeof GREY_COLORS)[number];

export type ColorsProps = Partial<{
  backgroundColor: Colors | GreyColors;
  textColor: Colors | GreyColors;
}>;

export const transformColorModifiers = makeTransform<ColorsProps>(
  props =>
    cx(props.className, {
      [`has-background-${props.backgroundColor}`]: props.backgroundColor,
      [`has-text-${props.textColor}`]: props.textColor,
    }),
  ["backgroundColor", "textColor"],
);
