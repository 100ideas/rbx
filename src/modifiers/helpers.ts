import { cx } from "emotion";

import { makeTransform } from "./utils";

export type HelpersProps = Partial<{
  clearfix: boolean;
  clipped: boolean;
  hidden: boolean;
  invisible: boolean;
  marginless: boolean;
  overlay: boolean;
  paddingless: boolean;
  pull: "right" | "left";
  radiusless: boolean;
  shadowless: boolean;
  unselectable: boolean;
}>;

export const transformHelpersModifiers = makeTransform<HelpersProps>(
  props =>
    cx(props.className, {
      "is-clearfix": props.clearfix,
      "is-clipped": props.clipped,
      "is-hidden": props.hidden,
      "is-invisible": props.invisible,
      "is-marginless": props.marginless,
      "is-overlay": props.overlay,
      "is-paddingless": props.paddingless,
      [`is-pulled-${props.pull}`]: props.pull,
      "is-radiusless": props.radiusless,
      "is-shadowless": props.shadowless,
      "is-unselectable": props.unselectable,
    }),
  [
    "clearfix",
    "clipped",
    "hidden",
    "invisible",
    "marginless",
    "overlay",
    "paddingless",
    "pull",
    "radiusless",
    "shadowless",
    "unselectable",
  ],
);
