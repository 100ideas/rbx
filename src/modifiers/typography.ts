import { cx } from "emotion";
import { ComponentProps } from "react";

export type TypographyProps = Partial<{
  textSize: 1 | 2 | 3 | 4 | 5 | 6;
  textAlignment: "centered" | "justified" | "left" | "right";
  textTransform: "capitalized" | "lowercase" | "uppercase";
  textWeight: "light" | "normal" | "semibold" | "bold";
  italic: boolean;
}>;

export function classNames(props: ComponentProps<any>) {
  return cx({
    [`has-text-${props.textAlignment}`]: props.textAlignment,
    [`has-text-weight-${props.textWeight}`]: props.textWeight,
    [`is-size-${props.textSize}`]: props.textSize,
    [`is-${props.textTransform}`]: props.textTransform,
    "is-italic": props.italic,
  });
}

export function clean({
  textWeight,
  textTransform,
  italic,
  textSize,
  textAlignment,
  ...props
}: ComponentProps<any>) {
  return props;
}
