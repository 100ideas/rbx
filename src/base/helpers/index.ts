import { Prefer } from "../../types";
import { TransformFunc } from "./types";

import { FloatHelpersProps, transformFloatHelpers } from "./float";
import { OtherHelpersProps, transformOtherHelpers } from "./other";
import { OverflowHelpersProps, transformOverflowHelpers } from "./overflow";
import { OverlayHelpersProps, transformOverlayHelpers } from "./overlay";
import {
  ResponsiveHelpersProps,
  transformResponsiveHelpers,
} from "./responsive";
import {
  transformTypographyHelpers,
  TypographyHelpersProps,
} from "./typography";
import {
  transformVisibilityHelpers,
  VisibilityHelpersProps,
} from "./visibility";

export { TransformFunc } from "./types";

/**
 * Union of helpers
 */
// tslint:disable-next-line: no-empty-interface
export interface HelpersPropsOverrides {}

export type HelpersProps = Prefer<
  HelpersPropsOverrides,
  FloatHelpersProps &
    OverflowHelpersProps &
    OverlayHelpersProps &
    TypographyHelpersProps &
    VisibilityHelpersProps &
    OtherHelpersProps &
    ResponsiveHelpersProps & { className?: string }
>;

export const combineTransformFunctions = <TTransformProps>(
  ...funcs: TransformFunc<any>[] // tslint:disable-line:no-any
): TransformFunc<TTransformProps> => (
  props,
  componentName,
  location = "prop",
) =>
  // tslint:disable-next-line:no-any
  funcs.reduce((acc, func) => func(acc, componentName, location) as any, props);

export const transformHelpers = combineTransformFunctions<HelpersProps>(
  transformFloatHelpers,
  transformOverflowHelpers,
  transformOverlayHelpers,
  transformTypographyHelpers,
  transformVisibilityHelpers,
  transformOtherHelpers,
  transformResponsiveHelpers,
);
