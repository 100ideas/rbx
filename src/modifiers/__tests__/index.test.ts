import { modify } from "..";
import { modify as colorModify } from "../color";
import { modify as helpersModify } from "../helpers";
// import { modify as responsiveModify } from "../responsives";
import { modify as typographyModify } from "../typography";

describe("Root modify", () => {
  test("should work on empty props", () => {
    expect(modify({})).toStrictEqual({});
  });

  test("should run all sub-modifiers", () => {
    expect(
      modify({
        backgroundColor: "info", // color
        clearfix: true, // helpers
        italic: true, // typography
        responsive: { textSize: 1 }, // responsive // todo!
      }),
    ).toMatchSnapshot();
  });

  test("should not clear out unknown props", () => {
    expect(modify({ unknowwn: true })).toMatchSnapshot();
  });
});

describe("Color modify", () => {
  test("should have class names applied", () => {
    expect(
      colorModify({
        backgroundColor: "info",
        textColor: "success",
      }),
    ).toMatchSnapshot();
  });
});

describe("Helpers modify", () => {
  test("should have class names applied", () => {
    expect(
      helpersModify({
        clearfix: true,
        clipped: true,
        hidden: true,
        invisible: true,
        marginless: true,
        overlay: true,
        paddingless: true,
        pull: "right",
        radiusless: true,
        shadowless: true,
        unselectable: true,
      }),
    ).toMatchSnapshot();
  });
});

// todo
// describe("Responsive modify", () => {
//   test("should have class names applied", () => {
//     expect(
//       responsiveModify({
//         responsive: []
//       }),
//     ).toMatchSnapshot();
//   });
// });

describe("Typography modify", () => {
  test("should have class names applied", () => {
    expect(
      typographyModify({
        italic: true,
        textAlignment: "centered",
        textSize: 1,
        textTransform: "capitalized",
        textWeight: "light",
      }),
    ).toMatchSnapshot();
  });
});

// todo

// test("Should have paddingless and clearfix classes", () => {
//   expect(
//     classNames({
//       clearfix: true,
//       paddingless: true,
//     }),
//   ).toMatchSnapshot();
// });

// test("Should have text helpers", () => {
//   expect(
//     classNames({
//       italic: true,
//       textAlignment: "centered",
//       textColor: "success",
//       textTransform: "uppercase",
//       textWeight: "bold",
//     }),
//   ).toMatchSnapshot();
// });

// test("Should have responsive modifier flex-tablet-only and block-widescreen", () => {
//   expect(
//     classNames({
//       responsive: {
//         tablet: {
//           display: {
//             only: true,
//             value: "flex",
//           },
//         },
//         widescreen: {
//           display: {
//             value: "block",
//           },
//         },
//       },
//     }),
//   ).toMatchSnapshot();
// });

// test("Should have hidden modifier tablet-only and widescreen", () => {
//   expect(
//     classNames({
//       responsive: {
//         tablet: {
//           hide: {
//             only: true,
//             value: true,
//           },
//         },
//         widescreen: {
//           hide: {
//             value: true,
//           },
//         },
//       },
//     }),
//   ).toMatchSnapshot();
// });

// test("Should have alignment responsive modifiers", () => {
//   expect(
//     classNames({
//       responsive: {
//         tablet: {
//           textAlignment: {
//             only: true,
//             value: "centered",
//           },
//         },
//         widescreen: {
//           textAlignment: {
//             value: "left",
//           },
//         },
//       },
//     }),
//   ).toMatchSnapshot();
// });

// test("Should have text color success", () => {
//   expect(
//     classNames({
//       textColor: "success",
//     }),
//   ).toMatchSnapshot();
// });

// test("Should have background color success", () => {
//   expect(
//     classNames({
//       backgroundColor: "success",
//     }),
//   ).toMatchSnapshot();
// });
