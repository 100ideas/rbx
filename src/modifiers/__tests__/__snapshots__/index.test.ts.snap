// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Color modify should have class names applied 1`] = `
Object {
  "className": "has-background-info has-text-success",
}
`;

exports[`Helpers modify should have class names applied 1`] = `
Object {
  "className": "is-clearfix is-clipped is-hidden is-invisible is-marginless is-overlay is-paddingless is-pulled-right is-radiusless is-shadowless is-unselectable",
}
`;

exports[`Root modify should not clear out unknown props 1`] = `
Object {
  "unknowwn": true,
}
`;

exports[`Root modify should run all sub-modifiers 1`] = `
Object {
  "className": "has-background-info is-clearfix is-italic",
}
`;

exports[`Typography modify should have class names applied 1`] = `
Object {
  "className": "has-text-centered has-text-weight-light is-size-1 is-capitalized is-italic",
}
`;
